ops/internal/deployer/deployer.go

```diff
@@ -0,0 +1,308 @@
+package deployer
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// OpDeployer manages the process of building a specific binary of op-deployer,
+// then shelling out to that binary for various cli commands
+type OpDeployer struct {
+	DeployerVersion    string
+	binaryPath         string
+	lgr                log.Logger
+	l1ContractsRelease string
+}
+
+// NewOpDeployer creates a new OpDeployer instance.
+func NewOpDeployer(lgr log.Logger, l1ContractsRelease string) (*OpDeployer, error) {
+	if l1ContractsRelease == "" {
+		return nil, fmt.Errorf("l1ContractsRelease cannot be empty")
+	}
+
+	opd := OpDeployer{
+		lgr:                lgr,
+		l1ContractsRelease: l1ContractsRelease,
+	}
+
+	err := opd.checkBinary()
+	if err != nil {
+		return nil, fmt.Errorf("failed binary check: %w", err)
+	}
+
+	return &opd, nil
+}
+
+// checkBinary checks if the op-deployer binary exists and is executable
+func (d *OpDeployer) checkBinary() error {
+	// Normalize the contracts string before lookup in versions.json
+	// 1. Remove tag:// prefix if present
+	// 2. Remove any -rc.X suffix for version matching
+	contractsKey := strings.TrimPrefix(d.l1ContractsRelease, "tag://")
+	rcSuffixRegex := regexp.MustCompile(`-rc\.[0-9]+$`)
+	contractsKey = rcSuffixRegex.ReplaceAllString(contractsKey, "")
+
+	// Read and parse versions.json
+	d.lgr.Info("Reading versions.json")
+	// Get the directory containing this file (deployer.go)
+	_, filename, _, _ := runtime.Caller(0)
+	deployerDir := filepath.Dir(filename)
+
+	versionsPath := filepath.Join(deployerDir, "versions.json")
+	versionsData, err := os.ReadFile(versionsPath)
+	if err != nil {
+		return fmt.Errorf("failed to read versions.json: %w", err)
+	}
+
+	var versions map[string]string
+	if err := json.Unmarshal(versionsData, &versions); err != nil {
+		return fmt.Errorf("failed to parse versions.json: %w", err)
+	}
+
+	deployerVersion, ok := versions[contractsKey]
+	if !ok {
+		return fmt.Errorf("no deployer version found for contracts: %s", contractsKey)
+	}
+	d.lgr.Info("Found deployer version", "version", deployerVersion)
+	d.DeployerVersion = deployerVersion
+
+	// Check if the op-deployer binary already exists
+	homeDir, err := os.UserHomeDir()
```

Comment:
Can we pass in the binary directory here as discussed rather than reading the user's home directory? This creates an implied API that we can make much clearer by passing in the directory on the command line.

Reply:
It will eventually be CLI invoked by other tooling, so what I'm suggesting we do here is update the deployer library to consume the binary directory as an arg so that we can pass it in via CLI later. The CI job that runs the report would specify the command line.